File: lab5/C++/src/lab5.cpp
#include <iostream>
#include <memory>
#include <vector>
#include "Unit.hpp"
#include "Staff.hpp"
#include "SupplyRequestDetail.hpp"
#include "SupplyResponse.hpp"
#include "Ammunition.hpp"
#include "Fuel.hpp"
#include "Vehicle.hpp"
#include "Weapon.hpp"

using namespace std;

int main()
{
    setlocale(LC_ALL, "Rus");
    srand(time(0)); // для генерации идентификаторов

    cout << "\nДемонстрация алгоритма снабжения\n" << endl;

    shared_ptr<Unit> army = make_shared<Unit>(100, "1-я Армия", UnitType::eArmy);
    shared_ptr<Unit> division = make_shared<Unit>(101, "2-я Мотострелковая Дивизия", UnitType::eDivision);

    // Склады для армии и дивизии
    shared_ptr<Warehouse> armyWarehouse = make_shared<Warehouse>(100, "Армейский склад", WarehouseType::eRear);
    shared_ptr<Warehouse> divisionWarehouse = make_shared<Warehouse>(101, "Дивизионный склад", WarehouseType::eField);

    // Наполнение склада дивизии
    vector<unique_ptr<Resource>> divisionResources;
    divisionResources.emplace_back(make_unique<Ammunition>(200, "Патроны 5.45мм", Caliber::e545mm, 100));
    divisionWarehouse->AddResources(divisionResources);

    // Наполнение склада дивизии
    vector<unique_ptr<Resource>> armyResources;
    armyResources.emplace_back(make_unique<Ammunition>(100, "Патроны 5.45мм", Caliber::e545mm, 2000));
    armyWarehouse->AddResources(armyResources);

    // Связываем подразделения со складами
    army->AssignWarehouse(armyWarehouse);
    division->AssignWarehouse(divisionWarehouse);

    // Связываем сами подразделения
    army->AddChildUnit(division);

    cout << "Создана иерархия: " << army->GetName() << " -> " << division->GetName() << endl;

    // Статические константыs

    cout << "MIN_COUNT: " << Equipment::MIN_COUNT << endl;
    cout << "MIN_QUANTITY: " << Resource::MIN_QUANTITY << endl;
    cout << "MIN_COMMANDER_RANK: " << (int)Unit::MIN_COMMANDER_RANK << endl;

    // Обработка исключений

    try {
        // Попытка создать невалидное снаряжение
        Weapon corrupted(rand(), "Оружие", Caliber::e545mm, -1);
    }
    catch (const std::invalid_argument& e) {
        cout << "Перехвачено исключение: " << e.what() << endl;
    }

    auto soldier1 = make_shared<Staff>(1, "Иванов Иван Иванович", Rank::ePrivate, Speciality::eInfantry);
    auto soldier2 = make_shared<Staff>(2, "Петров Петр Сергеевич", Rank::eSergeant, Speciality::eMachineGunner);
    auto soldier3 = make_shared<Staff>(3, "Иванов Алексей Владимирович", Rank::eCaptain, Speciality::eTanker);

    division->AddSoldier(soldier1);
    division->AddSoldier(soldier2);
    division->AddSoldier(soldier3);

    // Поиск по имени
    string searchQuery = "Иванов";
    auto foundSoldiers = division->FindByName(searchQuery);
    for (auto& weakSoldier : foundSoldiers) {
        if (auto soldier = weakSoldier.lock()) {
            cout << "Найден: " << *soldier << endl;
        }
    }

    // Перегрузка операторов сравнения

    cout << "солдат1 < солдат2: " << (*soldier1 < *soldier2) << endl;
    cout << "солдат2 > солдат1: " << (*soldier2 > *soldier3) << endl;

    // Перегрузка арифметических операторов

    SupplyRequestDetail testDetail(999, SupplyType::Ammunition, 100);
    testDetail.WithCaliber(Caliber::e545mm);
    cout << "Исходно: " << testDetail << endl;
    testDetail += 50;
    cout << "После += 50: " << testDetail << endl;
    testDetail -= 30;
    cout << "После -= 30: " << testDetail << endl;

    // Логика снабжения
    cout << "Дивизия запрашивает 2200 патронов (на её складе только 100, на армейском 2000):" << endl;

    SupplyRequestBuilder requestBuilder;
    unique_ptr<SupplyRequest> request = requestBuilder.WithAmmunition(Caliber::e545mm, 2200).Create();
    SupplyResponse response = division->MakeSupplyRequest(*request);

    cout << "Результат:" << endl;
    cout << "  Статус: " << SupplyResponse::StatusToString(response.GetStatus()) << endl;
    cout << "  Комментарий: " << response.GetComment() << endl;

    // Статические методы

    cout << "Ранг капитана: " << Staff::RankToString(Rank::eCaptain) << endl;
    cout << "Специальность танкиста: " << Staff::SpecialityToString(Speciality::eTanker) << endl;
}

// Дружественная функция перегрузки оператора << для SupplyRequestDetail

std::ostream& operator << (std::ostream& stream, const SupplyRequestDetail& detail) {
    stream << detail.ToString();
    return stream;
}

// Дружественная функция перегрузки оператора << для Staff
std::ostream& operator <<(std::ostream& stream, const Staff& staff) {
    stream << staff.ToString();
    return stream;
}

File: lab5/Java/src/Lab5.java

import enums.Caliber;
import enums.Rank;
import enums.Speciality;
import enums.SupplyType;
import enums.UnitType;
import enums.WarehouseType;
import static java.lang.System.out;
import java.util.ArrayList;
import java.util.Arrays;
import request.SupplyRequestDetail;
import request.SupplyResponse;
import unit.Staff;
import unit.Unit;
import warehouse.Warehouse;
import warehouse.equipments.Weapon;
import warehouse.resources.Ammunition;

class Lab5 {
    public static void main(String[] args) {

        Staff[] staff = new Staff[]
        {
             new Staff(1, "Иванов Иван Иванович", Rank.PRIVATE, Speciality.MEDIC),
             new Staff(2, "Морозов Константин Алексеевич", Rank.PRIVATE, Speciality.ENGINEER),
             new Staff(3, "Васильев Антон Владимирович", Rank.PRIVATE, Speciality.TANKER),
             new Staff(4, "Кузьмин Вячеслав Александрович", Rank.ARMY_GENERAL, Speciality.TANKER),
        };

        Unit army = new Unit(1, "1-я Армия", UnitType.ARMY);
        Unit division = new Unit(2, "1-я Мотострелковая дивизия", UnitType.DIVISION);
        Unit brigade = new Unit(3, "1-я Мотострелковая бригада", UnitType.BRIGADE);
        brigade.addSoldiers(new ArrayList<>(Arrays.asList(staff)));

        var searchResult = brigade.findByName("антон");
        out.println(String.format("Результаты поиска солдатов с именем \"Антон\": %s", searchResult));

        // Попытка создать снаряжение с некорректным колличеством

        try{
            Weapon corrupted = new Weapon(1, "Самоходная артиллерийская установка 2С1 «Гвоздика»", Caliber.C_122MM, 0);
        }
        catch(IllegalArgumentException ex) { out.println(String.format("Обработано исключение при создании снаряжения: %s", ex.getMessage())); }
        finally { out.println("Выполнен finally блок"); }

        army.assignCommander(staff[3]);
        
        Warehouse armyWarehouse = new Warehouse(1, "Армейский склад", WarehouseType.REAR);
        Warehouse divisionWarehouse = new Warehouse(2, "Дивизионный склад", WarehouseType.FIELD);


        armyWarehouse.addEquipment(new Weapon(1, "Самоходная артиллерийская установка 2С1 «Гвоздика»", Caliber.C_122MM, 20));
        armyWarehouse.addResource(new Ammunition(1, "Патроны 5.45мм", Caliber.C_545MM, 2000));
        divisionWarehouse.addResource(new Ammunition(1, "Патроны 5.45мм", Caliber.C_545MM, 100));

        army.assignWarehouse(armyWarehouse);
        division.assignWarehouse(divisionWarehouse);

        army.addChildUnit(division);
        division.addChildUnit(brigade);

        out.println(String.format("Создана иерархия:"));


        printTree(army, "");


        out.println("\nБригада запрашивает 2200 патронов и 25 артиллерийских установок калибра 122мм (на складе дивизии только 100 патронов, на армейском 2000 патронов и 20 артиллерийских установок):");


        SupplyRequestDetail[] details = new SupplyRequestDetail[]{
            new SupplyRequestDetail(1, SupplyType.AMMUNITION, 2200).withCaliber(Caliber.C_545MM),
            new SupplyRequestDetail(2, SupplyType.WEAPON, 25).withCaliber(Caliber.C_122MM)
        };

        var request = brigade.createRequest(new ArrayList<>(Arrays.asList(details)));
        SupplyResponse response = division.makeSupplyRequest(request);
        out.println("\nРезультат\n");
        out.println("Статус: "+SupplyResponse.statusToString(response.getStatus()));
        out.println("Комментарий: " + response.getComment());
    }

    public static void printTree(Unit parent, String tabs){
        
        out.println(tabs+parent.getName());

        
        for(var soldier : parent.getPersonnel()){
            out.println(tabs+'\t'+soldier.toString());
        }

        if(!parent.getChildren().isEmpty()){
            for(var child : parent.getChildren()){
                printTree(child, tabs+'\t');
            }
        }        
    }
}